README.md
[![Build status](https://ci.appveyor.com/api/projects/status/njsnuf2n3fiitkdm?svg=true)](https://ci.appveyor.com/project/Atari2/pixi)
[![CMake](https://github.com/Atari2/Pixi/actions/workflows/cmake.yml/badge.svg)](https://github.com/Atari2/Pixi/actions/workflows/cmake.yml)

PIXI is a tool to insert custom sprites into Super Mario World ROMs.

It assembles using Asar, meaning TRASM sprites are not supported.

Release Windows builds can be found [here](https://www.atarismwc.com/modern_pixi_releases/pixi.zip)
# README Contents:
- The List File
    - Normal Sprites
    - Per-Level Sprites
    - Other sprite types (cluster/extended)

- Sprite Insertion
    - Opening pixi.exe
    - Using the Command Prompt

- New Additions and Changes
    - Custom status handling and extra property bytes
    - Softcoding
    - Per-Level Sprites
    - SA-1 Detection and Default Labels
    - CFG Files and the new CFG Editor
    - JSON Files
    - Shared Routines
    - Header Files
    - Extra Bytes
    - Extend PIXI (extra defines and hijacks)

- Common Errors
    - JMP (label,x) or JSR (label,x)
    - JMP $xxxxxx
    - Faulty Math or Wrong Register size (inserts fine but crashes in-game)
    - incsrc/incbin file not found






# The List File
## Normal Sprites

To let the tool know which sprites you want to insert in which sprite slots, you have to open the file
called "list.txt" or create a new one. In it, you define the sprites you want to insert into your ROM and the sprite
number you want to give them. The format is as follows:

```
id SPACE cfg_file
```

For example, a "list.txt" that should insert Blue.cfg to sprite number 00, Green.cfg to sprite number 01
and Shooter.cfg to C0 with Generator.cfg at D0 would look as such:

```
    00 Blue.cfg
    01 Green.cfg
    C0 Shooter.cfg
    D0 Generator.cfg
```

Keep in mind that the areas for sprites (00-BF), shooters (C0-CF) and generators (D0-FF) are fixed.
If you want to insert a shooter, it has to be in the range of C0 to CF due to the sprite type's different coding.
    
## Per-Level Sprites (has to be enabled with -pl)

The slots B0 to BF are special, in that if you assign a level to them, they will become per-level sprites. The sprite will only use
the sprite slot between B0 and BF in that one specified level. Meaning you can assign sprite slot B0 of level 105
to a Thwomp and B0 of level 106 to a Hammer Bro if you wanted. Keep in mind this only holds true for slots B0 to BF.
Do note that per-level sprites support at max 4 extra bytes.
The format for per-level sprites looks as follows:

```
    level COLON id SPACE cfg_file
```

Here is an example of a "list.txt" file that inserts Blue.cfg and Green.cfg in the same slots as earlier,
and then inserts Red.cfg and Yellow.cfg only for level 105, while Boo.cfg uses the same slot as Red.cfg but in level 106:

```
    00 Blue.cfg
    01 Green.cfg

    105:B0 Red.cfg
    105:B1 Yellow.cfg
    106:B0 Boo.cfg

    or 

    B0 Blue.cfg 
    105:B0 Red.cfg
```
Note that the above is still perfectly valid, sprite B0 will behave like Blue.asm in any level except for 105, where it will take Red.asm properties and code instead.
This is because since Pixi 1.3, slots B0-BF are not exclusive to per-level sprites anymore but they can be used by normal sprites aswell instead

## Other sprite types
PIXI also has the ability to insert other types of sprites, such as cluster or extended sprites.
To insert these other types, you just have to change the list type within your list file. This is simply done by a
type of headline with the all caps type followed by a colon. Valid headlines are: "SPRITE:" (default), "EXTENDED:",
"CLUSTER:", all without quotes. You probably won't need the SPRITE: header, since it's the default but it's there
anyway.
After that header, you can proceed to place sprites just like before, except they are taken from their respective
directories. An example:

```
    00 Blue.cfg
    01 Green.cfg

    CLUSTER:
    00 flower.asm
    01 fish.asm

    EXTENDED:
    00 hammer.asm
```

Note that cluster and extended sprites use the .asm extension, while normal sprites have .cfg.
Also keep in mind that shooters and generators are part of the SPRITE: group and are seperated by their slot.

	
# Sprite Insertion
## Opening pixi.exe
When opening pixi.exe by double clicking it, a window will open asking you for the location of your Super Mario World ROM.
If the ROM is in the same folder as PIXI, you can simply type in its name, for example "My Hack.smc" and hit enter.
Otherwise, you can use your mouse to drag and drop your ROM into PIXI's window and then hit enter to insert all sprites
specified in the "list.txt" sprite list file explained above.

## Using the Command Prompt
This option is for slightly more advanced users. To open a command prompt window, hold shift and right click the folder PIXI
is in, then select "Open command window here".

The tool assumes a lot of default paths and files. You can change them when calling the tool from
the command line interface by typing them as in the example below.
```
    Usage: pixi <options> <ROM>
    Options are:
    -d              Enable debug output
    -out <filename>	This can be placed after "-d" and with it you can define a file to redirect the debug output instead of the prompt. This flag only works if the debug output flag is enabled.
    -k              Keep debug files
    -l  <listpath>  Specify a custom list file (Default: list.txt)
    -pl				Per level sprites - will insert perlevel sprite code
    -npl            Same as the current default, no sprite per level will be inserted, left dangling for compatibility reasons
    -d255spl		disables 255 sprite per level support (won't do the 1938 remap)
    -w              Enable asar warnings check, recommended to use when developing sprites

    -a  <asm>       Specify a custom asm directory (Default asm/)
    
    -sp <sprites>   Specify a custom sprites directory (Default sprites/)
    -sh <shooters>  Specify a custom shooters directory (Default shooters/)
    -g  <generators>        Specify a custom generators directory (Default generators/)
    -e  <extended>  Specify a custom extended sprites directory (Default extended/)
    -c  <cluster>   Specify a custom cluster sprites directory (Default cluster/)

    -r  <sharedpath>        Specify a shared routine directory (Default routines/)
    -nr <number>			Specify a maximum number of shared routines (Default is 100, maximum is 310)
    
    -ssc <append ssc>       Specify ssc file to be copied into <romname>.ssc
    -mwt <append mwt>       Specify mwt file to be copied into <romname>.mwt
    -mw2 <append mw2>       Specify mw2 file to be copied into <romname>.mw2, the provided file is assumed to have 0x00 first byte sprite header and the 0xFF end byte
    -s16 <base s16>         Specify s16 file to be used as a base for <romname>.s16
                            Do not use <romname>.xxx as an argument as the file will be overwriten
    
    -ext-off 		Turns off extmod logging
    -lm-handle <lm_handle_code>		Special command line to be used only within LM's custom user toolbar file. Available only on Windows.
```
MeiMei: meimei is an embedded tool pixi uses to fix sprite data for levels when sprite data size is changed for sprites already in use. That happens when you have a level that already uses a certain sprite and you change the amount of extra bytes said sprite uses.
```
    Options are:
    -meimei-off		Shuts down MeiMei completely
    -meimei-a		Enables always remap sprite data
    -meimei-k		Enables keep temp patches files
    -meimei-d		Enables debug for MeiMei patches - I recommend you turn this on if you wanna know what levels were edited and how.
```
They are all still considered pixi options by the way, so you would them the as any of ther other options above the MeiMei section.

## Examples:
    
`pixi.exe -l differentlistfile.txt rom.smc` -> will use "differentlistfile.txt" instead of "list.txt"

`pixi.exe -d -k -l differentlistfile.txt rom.smc`	-> will print debug output, keep debug files and use "differentlistfile.txt"
    
    
    
    
# New Additions and Changes
If you are used to using Romi's SpriteTool, here is a quick rundown of everything new added in PIXI:

## Custom status handling and extra property bytes:
As most people know, Pixi relies on 2 print statements to tell the game what code to run in which state of the sprite.
Most importantly, state 08 will run whatever code is under the "MAIN" print statement and state 01 will run whatever code is under the "INIT" print statement.

All the other states will run the corresponding vanilla code, however, some bits in !extra_prop_2 can be set to activate certain functions. Setting bit 7 of that byte will make the sprite run its MAIN code in any state and it won't run the vanilla code, setting bit 6 will make it run both vanilla code and the custom MAIN.

Since Pixi 1.2.16 you can have more control over other states that are not 08 and 01 by using new print statements crafted just for the occasion, valid print statements FOR NORMAL CUSTOM SPRITES are:

- `print "CARRIABLE", pc`  which will run in state 09
- `print "KICKED", pc`  which will run in state 0A
- `print "CARRIED", pc` which will run in state 0B
- `print "MOUTH", pc` which will run in state 07
- `print "GOAL", pc` which will run in state 0C

Note that while using these print statements, the data bank will be automatically set, so you don't need to manually set it like for MAIN or INIT.

Please be aware that the use of these labels completely and totally overrides ANY vanilla code that would run in the respective states (unless you set the aforementioned bits in the property bytes), 
so if you use them you have to code all of the wanted behaviors yourself, this is done on purpose so the code has complete control and they won't have unwanted side-effects due to vanilla code.

You can find the code that vanilla rom uses to handle those states at the following rom locations, you can use those to see how to do implement whatever you feel like vanilla gave you and you're missing now, maybe even better than how the original game did things:
- $01953C for carriable/stunned
- $019913 for kicked
- $019F71 for carried
- $018157 for goal tape (only activates when the sprite has "turn into a powerup at goal tape" bit on.

Fun fact, the game just returns when in Yoshi's Mouth so you can do anything you want here. Be aware that "MOUTH" activates only when the sprite is set to stay in Yoshi's mouth.


Resources/changelog.txt
Changelog
Version 1.32 (January 23, 2021):
  - (Atari2.0) Fixes spritetool_clean.asm not being in the asm/ folder.
  - (Atari2.0) Now permits nested shared routine calls, fixing https://github.com/JackTheSpades/SpriteToolSuperDelux/issues/28
  - (Atari2.0) Updated asar.dll to 1.80, fixes https://github.com/JackTheSpades/SpriteToolSuperDelux/issues/35
  - (Atari2.0) Updated and fixed make, Makefile and .gitignore
  - (Atari2.0) Added escaping to ! in filepaths, fixes https://github.com/JackTheSpades/SpriteToolSuperDelux/issues/16
  - (Atari2.0) Added autoclean to all sprite pointers, fixes https://github.com/JackTheSpades/SpriteToolSuperDelux/issues/34
  - (Atari2.0) Fixed various compiler warnings, swapped fread arguments causing an error when writing extmod file.
  - (Atari2.0) Fixed snestopc not working correctly on SA-1 roms bigger than 2 mbs
  - (Atari2.0) Greatly refactored and reworked MeiMei, embedded both ROM classes in one
  - (Atari2.0) Made Pixi work on +6mb ROMs correctly
  - (Atari2.0) Now can use -nr to specify max number of shared routines
  - (Atari2.0) Fixed various memory leaks, along with a couple of asar warnings related to relative paths being used.
  - (Atari2.0) You can now use comments in your list.txt file
  - (Atari2.0) Completely rewritten populate_sprite_list and relative_directory to fix various bugs related to pathing.

Version 1.31 (November 4, 2020):
  - (Atari2.0, randomdude999) Fully implemented Romi's SpriteTool legacy cleanup (note that this doesn't mean that now Pixi can coexist in the same rom as SpriteTool but it means that at least SpriteTool roms can be easily upgraded to Pixi without breaking massively as long as you do it properly). ASM file used for cleanup provided by randomdude999.
  - (Atari2.0) Fixed an issue where upon upgrading from 1.2.x to 1.3, Pixi would completely destroy overworld layer 2 events.
  - (Atari2.0) Upped limit for .s16 files from 0x10 pages to 0x38 pages
  - (Atari2.0) Added extmod support (for specifics read the Lunar Magic help file in the section "Restore Options"), also added a "-ext-off" command line option to turn off extmod logging.
  - (Atari2.0) Finished implementing -mw2 and -mwt command line options, since they were not completed.
  - (Atari2.0) Fixed a bunch of compilation warnings.
  - (Atari2.0) Added a command line option (-lm-handle) strictly to enable the use of Pixi in the custom user toolbar of Lunar Magic, check LM's help file for more details. Usage is: "-lm-handle <LM_VERIFICATION_HANDLE>", this command line option can't be used out of LM's custom user toolbars. Keep in mind that this is Windows only.
  - (Atari2.0) Added an icon to the executable file.

Version 1.3 (September 19, 2020):
  - (Atari2.0) Added a lot of utility routines (CapeContact, CheckForContactAlternate, SetPlayerClippingAlternate, SetSpriteClippingAlternate, LoseYoshi, FinishOAMWrite, FireballContact) most of these were taken from GIEPY. Credit to respective authors in the asm files.
  - (Atari2.0) Fixed a bug regarding cape interaction with extended sprites by adding a print "CAPE" statement, along with an ExtendedCapeClipping routine
  - (Atari2.0) Added a SpawnSpriteSafe routine to fix a problem related to how the vanilla game handles OAM in certain sprite slots.
  - (Atari2.0) Optimized Random a little bit
  - (Akaginite, randomdude999 and Atari2.0) Fixed, reworked and merged the per-level sprites update. Now you can use B0-BF as normal sprite slots aswell, reduced insert size
  - (Atari2.0) Added print statements for states that are not 8, more details in README.txt
  - (Atari2.0) Fixed various bugs related to per-level sprites in sprite.cpp
  - (Atari2.0) Added a warning in case the user is inserting Pixi while having the VRAM patch from LM disabled.
  - (Alcaro) Updated asardll.c to not give compiler warnings about strict aliasing and a missing cast
  - (randomdude999) Fixed compilation on Linux, added basic makefile
  - (lx5, Atari2.0) Updated CFG Editor and Pixi to support json files with more than 4 extra bytes for custom collections.
  - (Atari2.0) Changed SA1 definitions to use absolute addressing instead of long addressing.
  - (Atari2.0) Added setting bit to disable Lunar Magic sprite count warning message when using 255 sprites per level.
  - (Atari2.0) Fixed per-level sprites displaying 9 extra bytes in the Lunar Magic window, forced to 4 (as a side effect of this, in fact, per level sprites now can only use up to 4 extra bytes).

Version 1.2.15 (July 10, 2020):
  - (Atari2.0) Removed unnecessary waiting (BRA $00/NOPs) in SA-1 version of various routines (Aiming, CircleX, CircleY, Random)
  - (Atari2.0) Added exception control for missing .json files. Now it will output exactly what json file was not found instead of printing the stacktrace.
  - (Atari2.0) Added a command line option "-out" that can be added after "-d" to redirect debug output to a file instead of the prompt/stdout.
  - (Atari2.0) Fixed trimming of filenames causing bug with trailing whitespaces in .cfg/.json names in list.txt
  - (Atari2.0) Added a warning in case the user is inserting Pixi into a rom without having modified any level in Lunar Magic, causing various bugs (like goal posts not working).
  

Pixi/Rom.cpp
#include "Rom.h"

Rom::Rom(std::string romname) : m_name(romname)
{
	m_data.from_file(fileopen(romname.c_str(), "rb"));
	m_header_offset = m_data.size() & 0x7FFF;
	m_size = m_data.size() - m_header_offset;
	if (data()[0x7fd5] == 0x23) {
		if (data()[0x7fd7] == 0x0D) {
			m_mapper = MapperType::FullSA1Rom;
		}
		else {
			m_mapper = MapperType::SA1Rom;
		}
	}
	else {
		m_mapper = MapperType::LoRom;
	}
	ErrorState::debug("Correctly instantiated rom \"{}\" with mapper {} and header offset {:X}\n", m_name, MapperToString(m_mapper), m_header_offset);
}

Rom& Rom::operator=(Rom&& other) noexcept
{
	m_data = std::move(other.m_data);
	m_name = std::move(other.m_name);
	m_header_offset = other.m_header_offset;
	m_mapper = other.m_mapper;
	m_size = other.m_size;
	return *this;
}

constexpr MapperType Rom::mapper()
{
	return m_mapper;
}

// returns a read only view of the headerless data of the rom
const ByteArrayView<uint8_t> Rom::data()
{
	return ByteArrayView(m_data, m_header_offset);
}

void Rom::write(size_t offset, const uint8_t* data, size_t len) {
	m_data.write_at(data, len, offset);
}

void Rom::write(size_t offset, const std::vector<uint8_t>& data) {
	m_data.write_at(data.data(), data.size(), offset);
}

void Rom::write_snes(size_t address, const uint8_t* data, size_t len) {
	m_data.write_at(data, len, snes_to_pc(address));
}

void Rom::write_snes(size_t address, const std::vector<uint8_t>& data) {
	m_data.write_at(data.data(), data.size(), snes_to_pc(address));
}

uint8_t Rom::at(size_t offset)
{
	return m_data[offset];
}

uint8_t Rom::at_snes(size_t address)
{
	return m_data[snes_to_pc(address)];
}

uint8_t Rom::read_byte(size_t offset) {
	return m_data[m_header_offset + offset];
}

uint16_t Rom::read_word(size_t offset) {
	offset += m_header_offset;
	return m_data[offset] | m_data[offset + 1] << 8;
}

int Rom::read_long(size_t offset) {
	offset += m_header_offset;
	return m_data[offset] | m_data[offset + 1] << 8 | m_data[offset + 2] << 16;
}

int Rom::pointer_at_snes(size_t address) {
	auto offset = snes_to_pc(address);
	return m_data[offset] << 16 | m_data[offset + 1] << 8 | m_data[offset + 2];
}

size_t Rom::pc_to_snes(size_t address, bool header) {
	if (header)
		address -= m_header_offset;

	if (mapper() == MapperType::LoRom) {
		return ((address << 1) & 0x7F0000) | (address & 0x7FFF) | 0x8000;
	}
	else if (mapper() == MapperType::SA1Rom) {
		for (int i = 0; i < 8; i++) {
			if (sa1banks[i] == (address & 0x700000)) {
				return 0x008000 | (i << 21) | ((address & 0x0F8000) << 1) | (address & 0x7FFF);
			}
		}
	}
	else if (mapper() == MapperType::FullSA1Rom) {
		if ((address & 0x400000) == 0x400000) {
			return address | 0xC00000;
		}
		if ((address & 0x600000) == 0x000000) {
			return ((address << 1) & 0x3F0000) | 0x8000 | (address & 0x7FFF);
		}
		if ((address & 0x600000) == 0x200000) {
			return 0x800000 | ((address << 1) & 0x3F0000) | 0x8000 | (address & 0x7FFF);
		}
	}
	return s::max();
}

size_t Rom::snes_to_pc(size_t address, bool header) {

	if (mapper() == MapperType::LoRom) {
		if ((address & 0xFE0000) == 0x7E0000 || (address & 0x408000) == 0x000000 || (address & 0x708000) == 0x700000)
			return s::max();
		address = (address & 0x7F0000) >> 1 | (address & 0x7FFF);
	}
	else if (mapper() == MapperType::SA1Rom) {
		if ((address & 0x408000) == 0x008000) {
			address = sa1banks[(address & 0xE00000) >> 21] | ((address & 0x1F0000) >> 1) | (address & 0x007FFF);
		}
		else if ((address & 0xC00000) == 0xC00000) {
			address = sa1banks[((address & 0x100000) >> 20) | ((address & 0x200000) >> 19)] | (address & 0x0FFFFF);
		}
		else {
			address = s::max();
		}
	}
	else if (mapper() == MapperType::FullSA1Rom) {
		if ((address & 0xC00000) == 0xC00000) {
			address = (address & 0x3FFFFF) | 0x400000;
		}
		else if ((address & 0xC00000) == 0x000000 || (address & 0xC00000) == 0x800000) {
			if ((address & 0x008000) == 0x000000)
				return s::max();
			address = (address & 0x800000) >> 2 | (address & 0x3F0000) >> 1 | (address & 0x7FFF);
		}
		else {
			return s::max();
		}
	}
	else {
		return s::max();
	}

	return address + (header ? m_header_offset : 0);
}

void Rom::clean(PixiConfig& cfg)
{
	if (!strncmp((char*)m_data.ptr_at(snes_to_pc(0x02FFE2)), "STSD", 4)) { // already installed load old tables

		std::string path = cfg.AsmDir + "_cleanup.asm";
		FILE* clean_patch = fileopen(path.c_str(), "w");

		auto version = at(snes_to_pc(0x02FFE6));
		auto flags = at(snes_to_pc(0x02FFE7));

		bool per_level_sprites_inserted = ((flags & 0x01) == 1) || (version < 2);

		// bit 0 = per level sprites inserted
		if (per_level_sprites_inserted) {
			// remove per level sprites
			// version 1.30+
			if (version >= 30) {
				fmt::print(clean_patch, ";Per-Level sprites\n");
				int level_table_address = pointer_snes(0x02FFF1).addr();
				if (level_table_address != 0xFFFFFF && level_table_address != 0x000000) {
					auto pls_addr = snes_to_pc(level_table_address);
					for (int level = 0; level < 0x0400; level += 2) {
						int pls_lv_addr = (m_data[pls_addr + level] + (m_data[pls_addr + level + 1] << 8));
						if (pls_lv_addr == 0)
							continue;
						pls_lv_addr = snes_to_pc(pls_lv_addr + level_table_address);
						for (int i = 0; i < 0x20; i += 2) {
							auto pls_data_addr = (m_data[pls_lv_addr + i] + (m_data[pls_lv_addr + i + 1] << 8));
							if (pls_data_addr == 0)
								continue;
							Pointer main_pointer = pointer_at_snes(pls_data_addr + level_table_address + 0x0B);
							if (main_pointer.addr() == 0xFFFFFF) {
								continue;
							}
							if (!main_pointer.is_empty()) {
								fmt::print(clean_patch, "autoclean ${:06X}\t;{:03X}:{:02X}\n", main_pointer.addr(), level >> 1,
									0xB0 + (i >> 1));
							}
						}
					}
				}
				// version 1.2x
			}
			else {
				for (int bank = 0; bank < 4; bank++) {
					int level_table_address = (m_data[snes_to_pc(0x02FFEA + bank)] << 16) + 0x8000;
					if (level_table_address == 0xFF8000)
						continue;
					fmt::print(clean_patch, ";Per Level sprites for levels {:03X} - {:03X}\n", (bank * 0x80),
						((bank + 1) * 0x80) - 1);
					for (int table_offset = 0x0B; table_offset < 0x8000; table_offset += 0x10) {
						Pointer main_pointer = pointer_snes(level_table_address + table_offset);
						if (main_pointer.addr() == 0xFFFFFF) {
							fmt::print(clean_patch,
								";Encountered pointer to 0xFFFFFF, assuming there to be no sprites to clean!\n");
							break;
						}
						if (!main_pointer.is_empty()) {
							fmt::print(clean_patch, "autoclean ${:06X}\n", main_pointer.addr());
						}
					}
					fmt::print(clean_patch, "\n");
				}
			}
		}

		// if per level sprites are inserted, we only have 0xF00 bytes of normal sprites
		// due to 10 bytes per sprite and B0-BF not being in the table.
		// but if version is 1.30 or higher, we have 0x1000 bytes.
		const int limit = version >= 30 ? 0x1000 : (per_level_sprites_inserted ? 0xF00 : 0x1000);

		// remove global sprites
		fmt::print(clean_patch, ";Global sprites: \n");
		int global_table_address = pointer_snes(0x02FFEE).addr();
		if (pointer_snes(global_table_address).addr() != 0xFFFFFF) {
			for (int table_offset = 0x08; table_offset < limit; table_offset += 0x10) {
				Pointer init_pointer = pointer_snes(global_table_address + table_offset);
				if (!init_pointer.is_empty()) {
					fmt::print(clean_patch, "autoclean ${:06X}\n", init_pointer.addr());
				}
				Pointer main_pointer = pointer_snes(global_table_address + table_offset + 3);
				if (!main_pointer.is_empty()) {
					fmt::print(clean_patch, "autoclean ${:06X}\n", main_pointer.addr());
				}
			}
		}

		// remove global sprites' custom pointers
		fmt::print(clean_patch, ";Global sprite custom pointers: \n");
		int pointer_table_address = pointer_snes(0x02FFFD).addr();
		if (pointer_table_address != 0xFFFFFF && pointer_snes(pointer_table_address).addr() != 0xFFFFFF) {
			for (int table_offset = 0; table_offset < 0x100 * 15; table_offset += 3) {
				Pointer ptr = pointer_snes(pointer_table_address + table_offset);
				if (!ptr.is_empty() && ptr.addr() != 0) {
					fmt::print(clean_patch, "autoclean ${:06X}\n", ptr.addr());
				}
			}
		}

		// shared routines
		fmt::print(clean_patch, "\n\n;Routines:\n");
		for (int i = 0; i < 100; i++) {
			int routine_pointer = pointer_snes(0x03E05C + i * 3).addr();
			if (routine_pointer != 0xFFFFFF) {
				fmt::print(clean_patch, "autoclean ${:06X}\n", routine_pointer);
				fmt::print(clean_patch, "\torg ${:06X}\n", 0x03E05C + i * 3);
				fmt::print(clean_patch, "\tdl $FFFFFF\n");
			}
		}

		// Version 1.01 stuff:
		if (version >= 1) {

			// remove cluster sprites
			fmt::print(clean_patch, "\n\n;Cluster:\n");
			int cluster_table = pointer_snes(0x00A68A).addr();
			if (cluster_table != 0x9C1498) // check with default/uninserted address
				for (int i = 0; i < Sprite::SPRITE_COUNT; i++) {
					Pointer cluster_pointer = pointer_snes(cluster_table + 3 * i);
					if (!cluster_pointer.is_empty())
						fmt::print(clean_patch, "autoclean ${:06X}\n", cluster_pointer.addr());
				}

			// remove extended sprites
			fmt::print(clean_patch, "\n\n;Extended:\n");
			int extended_table = pointer_snes(0x029B1F).addr();
			if (extended_table != 0x176FBC) // check with default/uninserted address
				for (int i = 0; i < Sprite::SPRITE_COUNT; i++) {
					Pointer extended_pointer = pointer_snes(extended_table + 3 * i);
					if (!extended_pointer.is_empty())
						fmt::print(clean_patch, "autoclean ${:06X}\n", extended_pointer.addr());
				}
		}
		// everything else is being cleaned by the main patch itself.
		fclose(clean_patch);
		patch(path, cfg);
	}
	else if (!strncmp((char*)m_data.ptr_at(snes_to_pc(pointer_snes(0x02A963 + 1).addr() - 3)), "MDK", 3)) {
		ErrorState::pixi_error("It looks like this ROM has been patched with Romi's SpriteTool, Pixi is not compatible with it, insertion aborted\n");
	}
}

void addIncSrcToFile(FILE* fp, const std::vector<std::string>& toInclude) {
	for (std::string const& incPath : toInclude) {
		fmt::print(fp, "incsrc \"{}\"\n", incPath);
	}
}

bool Rom::patch(Sprite& spr, const std::vector<std::string>& extraDefines, PixiConfig& cfg) {
	std::string escapedDir = escapeDefines(spr.directory);
	std::string escapedAsmFile = escapeDefines(spr.asm_file);
	std::string escapedAsmDir = escapeDefines(cfg.AsmDir);
	FILE* sprite_patch = fileopen(Sprite::TEMP_SPR_FILE, "w");
	fmt::print(sprite_patch, "namespace nested on\n");
	fmt::print(sprite_patch, "incsrc \"{}sa1def.asm\"\n", escapedAsmDir);
	addIncSrcToFile(sprite_patch, extraDefines);
	fmt::print(sprite_patch, "incsrc \"shared.asm\"\n");
	fmt::print(sprite_patch, "incsrc \"{}_header.asm\"\n", escapedDir);
	fmt::print(sprite_patch, "freecode cleaned\n");
	fmt::print(sprite_patch, "warnings push\n");
	fmt::print(sprite_patch, "warnings disable w1005\n");
	fmt::print(sprite_patch, "SPRITE_ENTRY_{}:\n", spr.number);
	fmt::print(sprite_patch, "\tincsrc \"{}\"\n", escapedAsmFile);
	fmt::print(sprite_patch, "warnings pull\n");
	fmt::print(sprite_patch, "namespace nested off\n");
	fclose(sprite_patch);

	bool retval = patch(Sprite::TEMP_SPR_FILE, cfg);
	std::map<std::string, int> ptr_map = {
		std::pair<std::string, int>("init", 0x018021),
		std::pair<std::string, int>("main", 0x018021),
		std::pair<std::string, int>("cape", 0x000000),
		std::pair<std::string, int>("mouth", 0x000000),
		std::pair<std::string, int>("kicked", 0x000000),
		std::pair<std::string, int>("carriable", 0x000000),
		std::pair<std::string, int>("carried", 0x000000),
		std::pair<std::string, int>("goal", 0x000000)
	};
	int print_count = 0;
	auto asar_prints = asar_getprints(&print_count);
	std::vector<std::string> prints{};
	prints.reserve(print_count);
	for (int i = 0; i < print_count; i++) {
		prints.push_back({ asar_prints[i] });
		trim(prints[i]);
	}
	if (cfg.Debug)
		fmt::print("{}\n", spr.asm_file);
	if (print_count > 2 && cfg.Debug)
		fmt::print("Prints:\n");

	for (const auto& print : prints) {
		if (!print.compare(0, 4, "INIT"))
			ptr_map["init"] = std::stol(print.substr(4), nullptr, 16);
		else if (!print.compare(0, 4, "MAIN"))
			ptr_map["main"] = std::stol(print.substr(4), nullptr, 16);
		else if (!print.compare(0, 4, "CAPE") && spr.sprite_type == 1)
			ptr_map["cape"] = std::stol(print.substr(4), nullptr, 16);
		else if (!print.compare(0, 9, "CARRIABLE") && spr.sprite_type == 0)
			ptr_map["carriable"] = std::stol(print.substr(9), nullptr, 16);
		else if (!print.compare(0, 7, "CARRIED") && spr.sprite_type == 0)
			ptr_map["carried"] = std::stol(print.substr(7), nullptr, 16);
		else if (!print.compare(0, 6, "KICKED") && spr.sprite_type == 0)
			ptr_map["kicked"] = std::stol(print.substr(6), nullptr, 16);
		else if (!print.compare(0, 5, "MOUTH") && spr.sprite_type == 0)
			ptr_map["mouth"] = std::stol(print.substr(5), nullptr, 16);
		else if (!print.compare(0, 4, "GOAL") && spr.sprite_type == 0)
			ptr_map["goal"] = std::stol(print.substr(4), nullptr, 16);
		else if (!print.compare(0, 4, "VERG")) {
			if (PixiConfig::VERSION < std::stol(print.substr(4), nullptr, 16)) {
				ErrorState::pixi_error("Version guard failed on {}\n", spr.asm_file);
			}
		}
		else if (cfg.Debug) {
			ErrorState::debug("\t{}\n", print);
		}
		spr.table.init = Pointer(ptr_map["init"]);
		spr.table.main = Pointer(ptr_map["main"]);
		if (spr.table.init.is_empty() && spr.table.main.is_empty())
			ErrorState::pixi_error("Sprite {} had neither INIT nor MAIN defined in its file, insertion has been aborted.", spr.asm_file);
		if (spr.sprite_type == 1)
			spr.extended_cape_ptr = Pointer(ptr_map["cape"]);
		else if (spr.sprite_type == 0) {
			spr.ptrs.carried() = Pointer(ptr_map["carried"]);
			spr.ptrs.carriable() = Pointer(ptr_map["carriable"]);
			spr.ptrs.kicked() = Pointer(ptr_map["kicked"]);
			spr.ptrs.mouth() = Pointer(ptr_map["mouth"]);
			spr.ptrs.goal() = Pointer(ptr_map["goal"]);
		}
	}
	if (cfg.Debug) {
		if (spr.sprite_type == 0)
			fmt::print("\tINIT: ${:06X}\n\tMAIN: ${:06X}\n"
				"\tCARRIABLE: ${:06X}\n\tCARRIED: ${:06X}\n\tKICKED: ${:06X}\n"
				"\tMOUTH: ${:06X}\n\tGOAL: ${:06X}"
				"\n__________________________________\n",
				spr.table.init.addr(), spr.table.main.addr(), spr.ptrs.carriable().addr(),
				spr.ptrs.carried().addr(), spr.ptrs.kicked().addr(), spr.ptrs.mouth().addr(), spr.ptrs.goal().addr());
		else if (spr.sprite_type == 1)
			fmt::print("\tINIT: ${:06X}\n\tMAIN: ${:06X}\n\tCAPE: ${:06X}"
				"\n__________________________________\n",
				spr.table.init.addr(), spr.table.main.addr(), spr.extended_cape_ptr.addr());
		else
			fmt::print("\tINIT: ${:06X}\n\tMAIN: ${:06X}\n"
				"\n__________________________________\n",
				spr.table.init.addr(), spr.table.main.addr());
	}
	return retval;
}

bool Rom::patch(const std::string& dir, const std::string& file, PixiConfig& cfg) {
	std::string path = dir + file;
	return patch(path, cfg);
}

bool Rom::patch(const std::string& file, PixiConfig& cfg) {
	std::string patch_path = std::filesystem::absolute(file).generic_string();
	if (!asar_patch(patch_path.c_str(), (char*)m_data.ptr_at(m_header_offset), MAX_ROM_SIZE, &size())) {
		ErrorState::debug("Failure. Try fetch errors:\n");
		int error_count;
		auto errors = asar_geterrors(&error_count);
		fmt::print("An error has been detected:\n");
		for (int i = 0; i < error_count; i++)
			fmt::print("{}\n", errors[i].fullerrdata);
		ErrorState::pixi_error("An error was encountered in asar\n");
	}
	int warn_count = 0;
	auto loc_warnings = asar_getwarnings(&warn_count);
	for (int i = 0; i < warn_count; i++)
		cfg.WarningList.push_back(loc_warnings[i].fullerrdata);
	ErrorState::debug("Patching for {} successful\n", file);
	return true;
}

Pointer Rom::pointer_snes(int address, int size, int bank)
{
	auto offset = snes_to_pc(address);
	auto res = (m_data[offset]) | (m_data[offset + 1] << 8) | (m_data[offset + 2] << 16) * (size - 2);
	res |= (bank << 16);
	return Pointer(res);
}

void Rom::close()
{
	ErrorState::debug("Writing to ROM, size: {:X} bytes\n", m_data.size());
	FILE* fp = fileopen(m_name.c_str(), "wb");
	size_t written = fwrite(m_data.start(), sizeof(uint8_t), m_data.size(), fp);
	assert(written == m_data.size());
	fclose(fp);
}

void Rom::run_checks()
{
	auto version = at_snes(0x02FFE2 + 4);
	if (version > PixiConfig::VERSION && version != 0xFF) {
		ErrorState::pixi_error("The ROM has been patched with a newer version of PIXI (1.{:02d}) already. \nThis is version 1.{:02d}\nPlease get a newer version.\n", version, PixiConfig::VERSION);
	}

	auto lm_edit_ptr = pointer_at_snes(0x06F624);
	if (lm_edit_ptr == 0xFFFFFF) {
		ErrorState::pixi_warning("You're inserting Pixi without having modified a level in Lunar Magic, this will cause bugs\nDo you "
			"want to abort insertion now [y/n]?\nIf you choose 'n', to fix the bugs just reapply Pixi after having "
			"modified a level\n");
		char c = (char)getchar();
		if (tolower(c) == 'y') {
			ErrorState::pixi_error("Insertion was stopped, press any button to exit...\n");
		}
		fflush(stdin);
	}

	auto vram_jump = at_snes(0x00F6E4);
	if (vram_jump != 0x5C) {
		ErrorState::pixi_error("You haven't installed the VRAM optimization patch in Lunar Magic, this will cause many features of "
			"Pixi to work incorrectly, insertion was aborted...\n");
	}
}
